### PR TITLE
Implement JIT serialization of ProcessGroup

### PR DESCRIPTION
**Pull request by gmagogsfm**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48538 Bind all of frontend
* **#48333 Implement JIT serialization of ProcessGroup**

### CODE REVIEW

File: test/distributed/test_jit_c10d.py [hunk 1]
```diff
@@ -107,3 +107,41 @@ def run_pg_nccl_alltoall(
             return work.result()
 
         run_pg_nccl_alltoall(nccl_pg, output, input)
+
+class C10dFrontendJitTest(TestCase):
+
```
Comment (wanchaol) [inline test/distributed/test_jit_c10d.py:112]: nit: space remove?

File: test/distributed/test_jit_c10d.py [hunk 2]
```diff
@@ -107,3 +107,41 @@ def run_pg_nccl_alltoall(
...
+        tcp_store = torch.classes.dist_c10d.TCPStore(addr, port, 1, True)
+
+        ProcessGroupNCCL1 = frontend1.newProcessGroupHelper(
+            self.world_size, self.rank, [], "NCCL", tcp_store, "test_process_group", 0)
+
+        ProcessGroupNCCL2 = frontend2.get_process_group_by_name("test_process_group")
+        self.assertEqual(ProcessGroupNCCL1, ProcessGroupNCCL2)
+
+class C10dProcessGroupSerialization(TestCase):
+
```
Comment (wanchaol) [inline test/distributed/test_jit_c10d.py:128]: ditto

**Review by wanchaol — APPROVED** [review]

File: test/distributed/test_jit_c10d.py (hunk 1 above)
Comment (gmagogsfm) [reply test/distributed/test_jit_c10d.py:124]: Done.

File: torch/csrc/distributed/c10d/init.cpp [hunk 3]
```diff
@@ -1529,6 +1543,22 @@ static const auto ProcessGroupNCCLOptionsTorchBind =
...
+              auto name =
+                  ::c10d::DistributedC10d::get()->getNameOfProcessGroup(self);
+              return std::vector<std::string>{name};
+            },
+            [](std::vector<std::string> state) {
+              TORCH_CHECK(
+                  state.size() == 1,
+                  "Expecting exactly 1 state when restoring ProcessGroupNCCL, got: ",
+                  state.size());
+              auto base_process_group = ::c10d::DistributedC10d::get()->getProcessGroupByName(state.front());
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/init.cpp:1559]: what if we only get a NullProcessGroup here? shouldn't this consider as not successfully reloaded as the previous ProcessGroup already "disappear"?

**Pull request closed without merging**